RESTCOMM-1032 : Client Password Proposal 1a

abdulazizali77 opened the pull request:
-Simplest system wide algorithm implementation
-No changes to DAO, we assume all saved 'passwords' are actually hashed A1 -> MD5(user:realm:plaintext_password)
-We retain A1 functionality in DigestAuthentication
-Only send single WWW-Auth with System wide algorithm which should be MD5

Not implemented:
-"algorithm-sess"
-other algorithms aside from MD5

@gvagenas @maria-farooq George or Maria, please have a look

deruelle commented on restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/util/DigestAuthentication.java:
```diff
@@ -46,31 +46,53 @@ private static String A1(final String algorithm, final String user, final String
     }
 
     private static String A2(final String method, final String uri, String body, final String qop) {
+        return A2("MD5", method, uri, body, qop);
```
@abdulazizali77 why do we hardcode to MD5 instead of using client-algorithm from the settings ?

deruelle commented on restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/util/DigestAuthentication.java:
```diff
@@ -46,31 +46,53 @@ private static String A1(final String algorithm, final String user, final String
     }
 
     private static String H(final String data) {
+        return H(data, "MD5");
```
@abdulazizali77 why do we hardcode to MD5 instead of using client-algorithm from the settings ?

deruelle commented on restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/util/DigestAuthentication.java:
```diff
@@ -80,11 +102,20 @@ private static String H(final String data) {
 
     private static String KD(final String secret, final String data) {
-        return H(secret + ":" + data);
+        return KD(secret, data, "MD5");
```
@abdulazizali77 why do we hardcode to MD5 instead of using client-algorithm from the settings ?

deruelle commented on restcomm/restcomm.telephony.api/src/main/java/org/restcomm/connect/telephony/api/util/CallControlHelper.java:
```diff
@@ -47,6 +47,10 @@
 public class CallControlHelper {
 
     static boolean permitted(final String authorization, final String method, DaoManager daoManager) {
+        return permitted(authorization, method, daoManager, "MD5");
```
@abdulazizali77 why do we hardcode to MD5 instead of using client-algorithm from the settings ?

deruelle commented on restcomm/restcomm.telephony.api/src/main/java/org/restcomm/connect/telephony/api/util/CallControlHelper.java:
```diff
@@ -77,23 +81,41 @@ static boolean permitted(final String authorization, final String method, DaoMan
      * @throws IOException
      */
     public static boolean checkAuthentication(SipServletRequest request, DaoManager storage) throws IOException {
+        return checkAuthentication(request, storage, "MD5");
```
@abdulazizali77 why do we hardcode to MD5 instead of using client-algorithm from the settings ?

deruelle commented:
@abdulazizali77 did you run the testsuite  against this branch ? Can you provide the results and see if there is any diffs with current master ?

Did you test against popular SIP clients and our own SDKs (Web, Android, iOS) ?

abdulazizali77 commented:
@deruelle Jean, thanks for reviewing. I have 
-removed unused functions using "MD5" as default
-cleaned up some qop parameter passing

I have run the testsuite and as expected many tests failed, because the the default 1234 passwords need to be hashed. I have yet to do this.

I have tested only with Jitsi and Linphone currently. What other clients should we try? 
I do not know how to go about testing with the SDKs yet, who should i ask?

abdulazizali77 commented:
It turns out SipUnit also need changes to support cnonce and nc. Ive logged a bug against it
https://telestax.atlassian.net/browse/CLOUDOPS-512

deruelle commented:
Thanks @abdulazizali77. 

Please use the Olympus WebRTC application for Web and Android Olympus/iOS Olympus. You can sync up with @atsakiridis on this.

If you can fix sipunit too that would be great so we can guarantee end to end testing.

abdulazizali77 commented:
@deruelle @gvagenas Jean, George
I tested testsuite locally with new SipUnit and mss-arquillian and updated testsuite changes, the failure results without the changes are similar.
Tested manually WebRTC locally results are fine. A grep `src/main/webapp/resources/js/mobicents-libs/jain-sip.js` shows that implementation of cnonce and nc in Auth are correct

Have not tested Android SDK yet
Have not tested iOS SDK yet. However grepping the sofialib which it is using shows correct sending of cnonce and nc `dependencies/sources/sofia-sip/libsofia-sip-ua/iptsec/auth_client.c:1090`

The latest commits in `pom.xml` are referencing invalid SipUnit and mss-arquillian versions though, so will have to be edited when testing locally

abdulazizali77 commented:
Testsuite results are comparable getting 30 failures 31 errors. 
```
Results :

Failed tests:
  AccountsEndpointClosingTest.removeAccountAndReleaseProvidedNumbers:119 Expected at least one request matc                                    hing: {
  "urlPattern" : "/nexmo/number/cancel/.*/.*/US/12223334444",
  "method" : "POST"
}
Requests received: [ {
  "url" : "/restcomm-rvd/services/notifications",
  "absoluteUrl" : "http://localhost:8089/restcomm-rvd/services/notifications",
  "method" : "POST",
  "headers" : {
    "Authorization" : "Basic QUNBMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDo3N2Y4YzEyY2M3YjhmODQyM2U1YzM4Yj                                    AzNTI0OTE2Ng==",
    "Content-Length" : "76",
    "Content-Type" : "application/json; charset=UTF-8",
    "Host" : "localhost:8089",
    "Connection" : "Keep-Alive",
    "User-Agent" : "Apache-HttpClient/4.5.2 (Java/1.7.0_79)",
    "Accept-Encoding" : "gzip,deflate"
  },
  "body" : "[{\"type\":\"accountClosed\",\"accountSid\":\"ACA2000000000000000000000000000000\"}]",
  "browserProxyRequest" : false,
  "loggedDate" : 1502704705046,
  "loggedDateString" : "2017-08-14 10:58:25"
} ]
  CorsRelaxTest.corsHeadersAreReturnedForAccount:75 expected:<http://testing.restcomm.com> but was:<null>
  LiveCallModificationAnswerDelayTest.holdCall:556 expected:<2> but was:<0>
  LiveCallModificationAnswerDelayTest.testTerminateDialForkCallWhileRinging_LCM_to_dial_branches:748
  LiveCallModificationTest.holdCall:555 expected:<2> but was:<0>
  LiveCallModificationTest.testTerminateDialForkCallWhileRinging_LCM_to_dial_branches:738 expected:<0> but                                     was:<1>
  LiveCallModificationTest.testTerminateDialForkCallWhileRinging_LCM_to_initial_call:822 expected:<0> but w                                    as:<1>
  LiveCallModificationTest.testTerminateDialForkCallWhileRinging_LCM_to_move_initial_call_to_hangup_rcml:91                                    5 expected:<0> but was:<1>
  UsageRecordsTest.getUsageRecordsAlltime:214
  UsageRecordsTest.getUsageRecordsDaily:107
  UsageRecordsTest.getUsageRecordsMonthly:152
  UsageRecordsTest.getUsageRecordsYearly:187
  IncomingPhoneNumbersEndpointTest.getIncomingPhoneNumbersList:204
  IncomingPhoneNumbersEndpointTest.getIncomingPhoneNumbersListUsingPageSize:233
  VoxboneIncomingPhoneNumbersEndpointTest.testPurchaseAndDeletePhoneNumberSuccess:182
  VoxboneIncomingPhoneNumbersEndpointTest.testPurchasePhoneNumberNoPhoneNumberFound:277
  CallLifecycleTest.testDialCancelBeforeDialingClientAliceAfterTrying:247
  CallLifecycleTest.testSubAccountClientDialsParentAccountNumber:1700
  ClientsDialTest.testDialForkClientWebRTCBob_And_AliceWithMultipleRegistrations:1228
  ClientsDialTest.testDialForkClient_AliceMultipleRegistrations_George:1149
  DialActionAnswerDelayTest.testDialActionInvalidCall:205 expected:<180> but was:<100>
  DialActionAnswerDelayTest.testDialActionInvalidCallCheckCallStatusCompleted:276
  DialActionTest.testDialActionInvalidCall:227
  DialActionTest.testDialActionInvalidCallCheckCallStatusCompleted:286
  DialRecordingS3UploadAnswerDelayTest.testDialClientAlice_AliceDisconnects:312
  DialRecordingS3UploadAnswerDelayTest.testDialClientAlice_BobDisconnects:232
  ImsClientsDialAnswerDelayTest.testWebRTCClientIncomingBusy:548
  TestDialVerbPartTwoAnswerDelay.testDialClientAliceWithRecordAndStatusCallbackForApp:438
  TestMgcpOperations.testDialHelloPlay:161
  UserAgentManagerTest.registerUserAgentWithExtraParamsAndOptionsPing:265
Tests in error:
  RvdProjectsMigratorWorkspaceMigratedTest.checkIncomingPhoneNumbers:93 » IllegalState
  RvdProjectsMigratorWorkspaceMixedTest.checkIncomingPhoneNumbers:131 » IllegalState
  RvdProjectsMigratorWorkspaceOriginalTest.checkIncomingPhoneNumbers:114 » IllegalState
org.restcomm.connect.testsuite.http.AccountsEndpointTest.testCloseAccountCheckClient(org.restcomm.connect.t                                    estsuite.http.AccountsEndpointTest)
  Run 1: AccountsEndpointTest.testCloseAccountCheckClient:443
  Run 2: AccountsEndpointTest.stopSipStack:119 » Runtime java.lang.ArrayIndexOutOfBound...

org.restcomm.connect.testsuite.http.AccountsEndpointTest.testCreateAccountCheckClientExisted(org.restcomm.c                                    onnect.testsuite.http.AccountsEndpointTest)
  Run 1: AccountsEndpointTest.testCreateAccountCheckClientExisted:367
  Run 2: AccountsEndpointTest.stopSipStack:119 » Runtime java.lang.ArrayIndexOutOfBound...

  AccountsEndpointTest.testCreateAccountWithJapaneseChars:177 NullPointer
org.restcomm.connect.testsuite.http.AccountsEndpointTest.testUpdateAccountCheckClient(org.restcomm.connect.                                    testsuite.http.AccountsEndpointTest)
  Run 1: AccountsEndpointTest.testUpdateAccountCheckClient:405
  Run 2: AccountsEndpointTest.stopSipStack:119 » Runtime java.lang.ArrayIndexOutOfBound...

  ReferTest.org.restcomm.connect.testsuite.telephony.ReferTest » Runtime Could n...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithRecordAndStatusCallback                                    ForApp(org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo)
  Run 1: TestDialVerbPartTwo.testDialClientAliceWithRecordAndStatusCallbackForApp:645 expected:<3> but was:                                    <4>
  Run 2: TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...

org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithRecordAndStatusCallback                                    ForAppForThreeCalls(org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo)
  Run 1: TestDialVerbPartTwo.testDialClientAliceWithRecordAndStatusCallbackForAppForThreeCalls:865
  Run 2: TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...

org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialConferenceNoDialActionNoSms(org.restco                                    mm.connect.testsuite.telephony.TestDialVerbPartTwo)
  Run 1: TestDialVerbPartTwo.testDialConferenceNoDialActionNoSms:508
  Run 2: TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...

  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testRecordWithActionAndStatusCallbackForAppWit                                    hBobSendsFinishKey(org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo)
  Run 1: TestDialVerbPartTwo.testRecordWithActionAndStatusCallbackForAppWithBobSendsFinishKey:1029
  Run 2: TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...

  TestDialVerbPartTwo.after:148 » NoSuchMethod java.util.concurrent.ConcurrentHa...

Tests run: 569, Failures: 30, Errors: 31, Skipped: 12

[ERROR] There are test failures.
```
This run is from https://mobicents.ci.cloudbees.com/job/RestComm-For-Specific-Branch-No-Test/13/

abdulazizali77 commented:
The Android SDK test pcaps are here 
https://telestax.atlassian.net/browse/RESTCOMM-1052

abdulazizali77 commented:
@deruelle @maria-farooq 
This implementation will be impacted by organization feature changes. We cannot have a global config for `realm` given that all the domain requests will be different.
I would expect that the `restcomm_clients` table will need to have an additional `realm` column that will contain the organization specific uri.
Any thoughts?

deruelle commented:
@abdulazizali77 given the REGISTER or INVITE request  is done against a given Request URI or To URL that contains the domain this can be used as a realm and identify the client in the Organization, I'm not sure you need to duplicate the domain information from the Organization table you can just get it from there no ?

abdulazizali77 commented:
@deruelle Jean, the password hash created and saved in the db is actually `hash(A1)` which is `hash(username:realm:password)`.
When a new client is added, we have to create the hash there and then so that we dont have to save the plaintext password in our systems (which was the original concern).
When we receive the `realm` from any message later, we cannot just recreate `hash(A1)` without having saved the plaintext password, which is why having a fixed `realm` value from the beginning matching any corresponding `realm`s sent in messages is integral.

We cant hash the plaintext password AND then expect to reconstruct `hash(A1)` without unhashing the hashed password first (which is untenable), `hash(username:realm:password)` ~= `hash(username:realm:hash(password))`

Other implementations (like Asterisk) actually saves both `hash(A1)` and plain text password in separate columns, to perhaps maintain backwards compatibility. But in such an implementation, this of course will pose a similar security concern.

deruelle commented:
@abdulazizali77 ok I think we are saying the same thing, when you save  `hash(A1)` you take the realm from the Organization the client is part of to generate the hash and save it to DB. When later on a message comes, we do the same thing to generate the hash and compare it to the new one saved in DB.

What I was saying that storing the realm domain in `restcomm_clients` seems redundant since we can find it in the Organization table the client is belonging to.

jaimecasero commented on restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/util/DigestAuthentication.java:
```diff
@@ -97,6 +127,28 @@ private static void validate(final String user, final String realm, final String
+    }
+
+    public static void main(String [] str){
+        String username="alice";
```
should we turn this into unit test? or just remove it...

maria-farooq replied:
we will remove it before merge

maria-farooq replied:
will add a unit test as well

jaimecasero commented on restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/entities/Client.java:
```diff
@@ -148,8 +149,14 @@ public Client setFriendlyName(final String friendlyName) {
+     * @return
+     */
+    public Client setPassword(final String login, final String password, final String realm) {
+        return new Client(sid, dateCreated, DateTime.now(), accountSid, apiVersion, friendlyName, login, DigestAuthentication.HA1(login, realm, password), status,
```
as mentionedin daily, i would prefer this kind of logic to be isolated in a Service class, and let the Entity as DTO container. The potential new service component could hold the logic of DigestAuthentication utility class as well. We may refactor this in a next iteration if scope in current Sprint is not appropiate.

jaimecasero commented on restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/util/DigestAuthentication.java:
```diff
@@ -97,6 +127,28 @@ private static void validate(final String user, final String realm, final String
+     * @return
+     */
+    public static String HA1(String username, String realm, String password){
+        String algorithm = RestcommConfiguration.getInstance().getMain().getClientAlgorithm();
```
could we pass the algorithm as parameter here to reduce dependency to RestCommConfiguration in util class?. This may facilitate a next iteration where algorithm is specified per Client, instead of systemWide setting...

maria-farooq replied:
removing dependency from RestCommConfiguration in utill class would mean to replicate it in all calling classes. for now its ok, it would make more sense when we wil support multiple algorithms
